### PR TITLE
transport: return unexpected EOF for partial reads

### LINKED ISSUE #9200: transport: return unexpected EOF for partial message headers
### Summary

`Stream.ReadMessageHeader` is intended to return `io.ErrUnexpectedEOF` when EOF occurs after only part of the 5-byte gRPC message header has been read. When the partial bytes and EOF arrive from separate recv buffer messages, the final read can return `n == 0, io.EOF`, so the method returns clean EOF even though earlier iterations already read part of the header.

### Why this matters

The error returned by `ReadMessageHeader` tells callers whether the stream ended cleanly or whether a gRPC message header was truncated. Returning `io.EOF` after partial header progress hides the fact that malformed or truncated message data was observed.

This makes truncated message-header handling less precise and can make debugging or higher-level error handling harder.

### Environment

**gRPC version:** current master at `d74861eb`

**Go version:** `go version go1.25.1 X:nodwarf5 linux/amd64`

**OS:** `Linux claude 6.16.5-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 04 Sep 2025 23:18:13 +0000 x86_64 GNU/Linux`

### Trigger

Read a gRPC message header where some bytes of the 5-byte header arrive first, and EOF arrives later as a separate recv buffer message.

### Expected behavior

Any EOF after partial header progress should be reported as `io.ErrUnexpectedEOF`.

### Actual behavior

Partial header data followed by EOF can be reported as `io.EOF`.

### Proposed fix

PR #9204 tracks cumulative bytes read while filling a gRPC message header. If EOF occurs after any partial progress, it returns `io.ErrUnexpectedEOF` even when the final read call itself returned zero bytes. The PR includes a regression test that splits a partial header and EOF across separate recv buffer messages.

### PR DESCRIPTION
**Pull request by buger**
Fixes #9200

Track cumulative bytes read while filling a gRPC message header or body. If EOF occurs after any partial progress, return `io.ErrUnexpectedEOF` even when the final read call itself returned zero bytes. Preserve plain `io.EOF` when EOF occurs before any data is read.

Regression tests split partial headers and bodies from EOF across separate receive-buffer messages. They verify that partial reads return `io.ErrUnexpectedEOF` while zero-progress reads retain `io.EOF`.

RELEASE NOTES:
* transport: Truncated gRPC message headers and bodies now return `io.ErrUnexpectedEOF` after partial data is received.

### CODE REVIEW

**Comment by buger** [comment]
Maintainer note: the `Validate PR` check is currently failing only because this PR has no `Type:` label. I don't have permission to add labels on this repository; `Type: Bug` seems appropriate for this fix.

File: internal/transport/transport_test.go [hunk 1]
```diff
@@ -3387,6 +3387,36 @@ func (s) TestReadMessageHeaderMultipleBuffers(t *testing.T) {
...
+	recvBuffer.put(recvMsg{buffer: make(mem.SliceBuffer, 3)})
+	recvBuffer.put(recvMsg{err: io.EOF})
+
+	if err := s.ReadMessageHeader(make([]byte, headerLen)); !errors.Is(err, io.ErrUnexpectedEOF) {
+		t.Fatalf("ReadMessageHeader() error = %v, want %v", err, io.ErrUnexpectedEOF)
+	}
+	if bytesRead != 3 {
+		t.Fatalf("bytesRead = %d, want 3", bytesRead)
+	}
+}
```
[the remark this answers is not in the document]
Comment (mbissa) [reply internal/transport/transport_test.go:3418]: Please make this change, it is a valid suggestion.

File: internal/transport/transport_test.go [hunk 2]
```diff
@@ -3387,6 +3387,36 @@ func (s) TestReadMessageHeaderMultipleBuffers(t *testing.T) {
...
+		},
+		windowHandler: &mockWindowUpdater{
+			f: func(i int) {
+				bytesRead += i
+			},
+		},
+	}
+
+	recvBuffer.put(recvMsg{buffer: make(mem.SliceBuffer, 3)})
+	recvBuffer.put(recvMsg{err: io.EOF})
```
Comment (mbissa) [inline internal/transport/transport_test.go:3410]: can we also have a separate test case for negative testing where we check only for recvMsg{err: io.EOF} and verify it is reported as EOF and not ErrUnexpectedEOF incorrectly.

**Comment by mbissa** [comment]
Thanks @buger for the contribution, can you please address the comments once?

**Comment by buger** [comment]
Thanks @mbissa, I updated this PR to address the review comments.

Changes applied:
- renamed the local `Stream` variable in `TestReadMessageHeaderPartialHeaderEOF` to avoid shadowing the receiver
- added a separate EOF-only negative test to verify `recvMsg{err: io.EOF}` is reported as `io.EOF` and not `io.ErrUnexpectedEOF`

Local validation passed:
```bash
go test ./internal/transport -run '^Test/ReadMessageHeader' -count=1
git diff --check
```

File: internal/transport/transport_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (buger) [reply internal/transport/transport_test.go:3418]: Done in 472254d2. The local `Stream` variable is now named `stream` throughout `TestReadMessageHeaderPartialHeaderEOF`.

File: internal/transport/transport_test.go (hunk 2 above)
Comment (buger) [reply internal/transport/transport_test.go:3410]: Done in 472254d2. I added `TestReadMessageHeaderEOF` with only `recvMsg{err: io.EOF}`; it asserts `errors.Is(err, io.EOF)` and explicitly rejects `io.ErrUnexpectedEOF`. `go test ./internal/transport -count=1` passes.

**Comment by buger** [comment]
@mbissa, I have now replied directly to both review threads and resolved them after re-verifying the requested changes in 472254d2. The full `internal/transport` package passes locally with `go test ./internal/transport -count=1`.

Could you please re-review when convenient and clear `Status: Requires Reporter Clarification`? The PR is currently also marked `stale`.

**Review by mbissa — APPROVED** [review]
LGTM, assigning to @easwars for second set of eyes.

File: internal/transport/transport_test.go [hunk 3]
```diff
@@ -3404,6 +3387,60 @@ func (s) TestReadMessageHeaderMultipleBuffers(t *testing.T) {
...
+	recvBuffer := &stream.buf
+	stream.trReader = transportReader{
+		reader: recvBufferReader{
+			recv: recvBuffer,
+		},
+		windowHandler: &mockWindowUpdater{
+			f: func(i int) {
+				bytesRead += i
+			},
+		},
```
Comment (easwars) [inline internal/transport/transport_test.go:3406]: Nit: Is this required. Can it be an empty one like in the next test? Or should it be a non-nil func like:
```golang
		windowHandler: &mockWindowUpdater{f: func(int) {}},
```

File: internal/transport/transport.go
Comment (easwars) [inline internal/transport/transport.go:428]: It looks like we have the same bug here as well. Any EOF in this block when `n` is still greater than `0` must be treated like an unexpected EOF.

**Review by easwars — COMMENTED** [review]
Apologies for the delay in reviewing this change.

File: internal/transport/transport_test.go (hunk 3 above)
Comment (buger) [reply internal/transport/transport_test.go]: Done in d4dc600d. I replaced the byte-accounting callback and assertion with the suggested non-nil no-op callback. An empty `mockWindowUpdater` is not safe in this test because the partial read invokes `updateWindow`, whose mock calls `f`; the no-op keeps the test focused without a nil-function panic.

File: internal/transport/transport.go
Comment (buger) [reply internal/transport/transport.go:530]: Done in d4dc600d. `Stream.read` now tracks cumulative bytes across iterations and converts EOF to `io.ErrUnexpectedEOF` after partial progress. I added `TestReadPartialMessageEOF`, which failed with plain EOF before the fix, plus `TestReadMessageEOF` to preserve the established plain-EOF behavior when no bytes were read. I used cumulative progress rather than `n > 0` alone because the latter changed zero-progress end-of-stream reads and broke existing transport tests. Focused tests, the focused race run, and `go test ./internal/transport -count=1` all pass.

**Comment by buger** [comment]
@easwars, I addressed both review comments and replied inline. The partial-message read now uses cumulative progress so EOF after prior bytes becomes `io.ErrUnexpectedEOF`, while zero-progress EOF remains `io.EOF`; focused tests cover both boundaries. I also updated the test window handler to the suggested non-nil no-op.

The branch is now current with `master`, and the complete CI matrix is green. Could you please re-review when convenient and clear `Status: Requires Reporter Clarification`?

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
Thanks for your contribution, @buger
